k-distribution/include/kframework/builtin/domains.md

```diff
@@ -2419,6 +2419,17 @@ finishes.
   syntax K ::= #logToFile(name: String, value: String) [function, functional, hook(IO.log), impure, returnsUnit, symbol]
 ```
 
+Terms can also be logged to standard error in _surface syntax_, rather than as
+KORE using `#trace`. This operator has similar semantics to `#logToFile` (i.e.
+it returns `.K`, but prints as an impure side effect). Note that calling
+`#trace` is equivalent to invoking the `kprint` tool for the first term that is
+logged, which requires re-parsing the underlying K definition. Subsequent calls
+do not incur this overhead again; the definition is cached.
+
+```k
+  syntax {Sort} K ::= #trace(value: Sort) [function, functional, hook(IO.traceTerm), impure, returnsUnit, symbol]
```

Comment:
I have to wonder if we might be better off changing the signature of this somewhat. People are probably going to end up pretty surprised if they try to trace a builtin sort, because it's not going to work. I would suggest we instead have two overloads, both of which have the same hook tied to them: one that takes a KItem and one that takes a K. The reason I suggest two is that if you just do K, then everything will get pretty-printed with a `~> .` at the end, but if you just do KItem, then you won't be able to pretty print k sequences. That being said, using the KItem one in most cases ought to be sufficient because it will automatically inject whatever sort you want into KItem and the injection will be pretty-printed invisibly, so you ought to get the same output as if you had pretty-printed a particular sort. Except that it will be able to handle builtin sorts.

Reply:
What's the best way to write that overloaded version? If I have:
```k
  syntax K ::= #trace(value: K)     [function, functional, hook(IO.traceTerm), impure, returnsUnit, symbol, klabel(#traceK)]
             | #trace(value: KItem) [function, functional, hook(IO.traceTerm), impure, returnsUnit, symbol, klabel(#traceKItem)]
```

then I get parsing ambiguities on the argument:
```
[Error] Inner Parser: Parsing ambiguity.
1: syntax K ::= "#trace" "(" K ")" [function, functional, hook(IO.traceTerm), impure, klabel(#traceK), returnsUnit, symbol]
    #traceK(`bar_*__TEST_Bar_Foo_Foo`(`foo(_)_TEST_Foo_Int`(#token("2","Int")),`foo(_)_TEST_Foo_Int`(#token("0","Int"))))
2: syntax K ::= "#trace" "(" KItem ")" [function, functional, hook(IO.traceTerm), impure, klabel(#traceKItem), returnsUnit, symbol]
    #traceKItem(`bar_*__TEST_Bar_Foo_Foo`(`foo(_)_TEST_Foo_Int`(#token("2","Int")),`foo(_)_TEST_Foo_Int`(#token("0","Int"))))
	Source(/home/bruce/code/k/k-distribution/tests/regression-new/trace/test.k)
	Location(13,20,13,47)
	13 |	    (#trace(F2) ~> #trace(bar foo(2) * foo(0)) ~> .K)
	   .	                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```